File: sections/31_sieve_primes_factorization/problems/c_divisor_queries/solution.py
import sys


def spf_table(n: int) -> list[int]:
    spf = list(range(n + 1))
    if n >= 1:
        spf[1] = 1
    for p in range(2, n + 1):
        if spf[p] == p and p * p <= n:
            for x in range(p * p, n + 1, p):
                if spf[x] == x:
                    spf[x] = p
    return spf


def divisors(x: int, spf: list[int]) -> int:
    ans = 1
    while x > 1:
        p = spf[x]
        exp = 0
        while x % p == 0:
            exp += 1
            x //= p
        ans *= exp + 1
    return ans


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    values = data[1 : 1 + q]
    spf = spf_table(max(values))
    print("\n".join(str(divisors(x, spf)) for x in values))


if __name__ == "__main__":
    main()

File: sections/31_sieve_primes_factorization/problems/c_divisor_queries/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<int> values(q);
    int mx = 1;
    for (int &x : values) {
        cin >> x;
        mx = max(mx, x);
    }
    vector<int> spf(mx + 1);
    iota(spf.begin(), spf.end(), 0);
    if (mx >= 1) spf[1] = 1;
    for (long long p = 2; p <= mx; p++) {
        if (spf[p] == p && p * p <= mx) {
            for (long long x = p * p; x <= mx; x += p) {
                if (spf[x] == x) spf[x] = p;
            }
        }
    }
    for (int x : values) {
        int ans = 1;
        while (x > 1) {
            int p = spf[x];
            int exp = 0;
            while (x % p == 0) {
                exp++;
                x /= p;
            }
            ans *= exp + 1;
        }
        cout << ans << '\n';
    }
    return 0;
}
